File: examples/ts/branch-merged-locals/BranchMergedLocals.runar.ts
import { StatefulSmartContract } from 'runar-lang';

/**
 * BranchMergedLocals -- regression fixture for the branch-merged-local
 * miscompilation reported privately on 2026-08-03.
 *
 * Two locals are initialised from state fields and the arms of an `if`
 * reassign DIFFERENT ones, then both feed the continuation. An `if` carries a
 * single value, so ANF lowering could only rewire post-branch references for
 * ONE merged local; with two, every later reference still named the dead
 * pre-branch binding, and stack lowering registered one stack-map slot for two
 * physical results and resolved every later operand one slot off. The script
 * compiled cleanly, passed the ANF-interpreter tests, and was rejected by the
 * real Script interpreter -- a permanently unspendable contract from idiomatic
 * source.
 *
 * Both arms now end with the same canonical result block, so the merged values
 * survive whichever branch runs. See
 * packages/runar-testing/src/__tests__/branch-merged-locals-vm.test.ts for the
 * real-Script-VM proof.
 */
class BranchMergedLocals extends StatefulSmartContract {
  a: bigint;
  b: bigint;

  constructor(a: bigint, b: bigint) {
    super(a, b);
    this.a = a;
    this.b = b;
  }

  public bid(amount: bigint, toFirst: bigint) {
    let na = this.a;
    let nb = this.b;
    if (toFirst > 0n) {
      na = amount;
    } else {
      nb = amount;
    }
    this.addOutput(1000n, na, nb);
  }
}

File: examples/go/branch-merged-locals/BranchMergedLocals.runar.go
package x

import runar "github.com/icellan/runar/packages/runar-go"

// BranchMergedLocals -- regression fixture for the branch-merged-local
// miscompilation reported privately on 2026-08-03: two locals initialised from
// state, the arms of an `if` reassigning DIFFERENT ones, both feeding the
// continuation. Post-branch references used to resolve to the dead pre-branch
// binding.
type BranchMergedLocals struct {
	runar.StatefulSmartContract

	A runar.Bigint
	B runar.Bigint
}

func (c *BranchMergedLocals) Bid(amount runar.Bigint, toFirst runar.Bigint) {
	na := c.A
	nb := c.B
	if toFirst > 0 {
		na = amount
	} else {
		nb = amount
	}
	c.AddOutput(1000, na, nb)
}
